python/hail/methods/misc.py

```diff
@@ -358,4 +369,185 @@ def window_by_locus(mt: MatrixTable, bp_window_size: int) -> MatrixTable:
+
+    >>> hl.locus_windows(ht, 1)
+    (array([0, 0, 2, 3, 3, 5]), array([2, 2, 3, 5, 5, 6]))
+
+    Windows with 1cm radius:
+
+    >>> hl.locus_windows(ht, 1.0, value_expr=ht.cm)
+    (array([0, 1, 1, 3, 3, 5]), array([1, 3, 3, 5, 5, 6]))
+
+    Notes
+    -----
+    The first key field of `locus_table` must be of type :py:data:`tlocus`, as
+    is true for variant-keyed datasets. The rows are then ordered by locus, i.e.
+    by contig and then by position within contig.
+
+    This function returns two 1-dimensional ndarrays of integers,
+    ``starts`` and ``stops``, each of size equal to the number of rows.
+
+    By default, for all indices ``i``, ``[starts[i], stops[i])`` is the maximal
+    range of row indices ``j`` such that ``contig[i] == contig[j]`` and
+    ``position[i] - radius <= position[j] <= position[i] + radius``.
+
+    Set `value_expr` to use a value other than position to define the windows.
+    This row-indexed numeric expression on `locus_table` must be non-missing,
+    as well as non-decreasing with respect to locus for each contig; otherwise
+    the function will fail.
+
+    The last example above uses centimorgan coordinates, so
+    ``[starts[i], stops[i])`` is the maximal range of row indices ``j`` such
+    that ``contig[i] == contig[j]`` and
+    ``cm[i] - radius <= cm[j] <= cm[i] + radius``.
+
+    Index ranges are start-inclusive and stop-exclusive. This function is
+    especially useful in conjunction with
+    :meth:`.BlockMatrix.sparsify_row_intervals`.
+
+    Parameters
+    ----------
+    locus_table : :class:`.Table`
+        Table with first key of type :py:data:`tlocus`.
+    radius: :obj:`int`
+        Radius of window for row values.
+    value_expr: :obj:`str`, optional
+        Row-indexed numeric expression on `locus_table` for the row value.
+        By default, the row value is given by the locus position.
+
+    Returns
+    -------
+    (:class:`ndarray` of :obj:`int64`, :class:`ndarray` of :obj:`int64`)
+        Tuple of start indices array and stop indices array.
+    """
+    if radius < 0:
+        raise ValueError(f'locus_windows: radius must be non-negative, found {radius}')
+
+    require_first_key_field_locus(locus_table, 'locus_windows/locus_table')
+    ht = locus_table
+
+    if value_expr is None:
+        ht = ht.key_by(ht.key[0]).select()
+        contigs_table = ht.group_by(ht.key[0].contig).aggregate(values=agg.collect(ht.key[0].position))
+    else:
+        # FIXME: remove once select_entries on a field is free
+        if value_expr in ht._fields_inverse:
+            field = ht._fields_inverse[value_expr]
+            ht = ht.select(field)
+            ht = ht.key_by(ht.key[0])
+        else:
+            field = Env.get_uid()
+            ht = ht.select(**{field: value_expr})
+        contigs_table = ht.group_by(ht.key[0].contig).aggregate(values=agg.collect(ht[field]))
+
+    contig_values = contigs_table.order_by(contigs_table.contig).values.collect()
```

Comment:
I think this will fail if the contigs have a different sort order than the reference genome order.